File: src/billit_mcp/tools/search_fetch.py
"""ChatGPT Apps SDK compatibility — ``search`` and ``fetch`` tools.

ChatGPT (and especially Deep Research) expects servers to expose two specific
tools with these exact signatures:

* ``search(query) -> { results: [{ id, title, url }] }``
* ``fetch(id)    -> { id, title, text, url, metadata? }``

We map both onto Billit's orders + parties surface so the model can browse
invoices and contacts without thinking about Billit's resource taxonomy.

References:
- https://developers.openai.com/apps-sdk/build/mcp-server
"""

from __future__ import annotations

import json
from typing import Any

from fastmcp import FastMCP

from ..client import BillitClient
from ._common import odata_params

_ORDER_PREFIX = "order:"
_PARTY_PREFIX = "party:"


def register(mcp: FastMCP, client: BillitClient) -> None:
    @mcp.tool(
        annotations={
            "title": "Search Billit",
            "readOnlyHint": True,
            "openWorldHint": True,
        }
    )
    async def search(query: str) -> dict[str, Any]:
        """Search invoices and customers/suppliers across Billit.

        Returns ChatGPT Apps SDK-compatible records:
        ``{ "results": [ { "id": "order:123", "title": "...", "url": "..." } ] }``
        """
        results: list[dict[str, str]] = []

        # Orders — first page, sorted newest first.
        orders_resp = await client.get(
            "orders",
            params=odata_params(top=10, full_text_search=query, orderby="OrderDate desc"),
        )
        for o in _items(orders_resp):
            oid = o.get("OrderID") or o.get("Id")
            if oid is None:
                continue
            title = (
                f"{o.get('OrderType', 'Order')} {o.get('OrderNumber', '')} — "
                f"{(o.get('Customer') or {}).get('Name', '')}"
            ).strip(" —")
            results.append(
                {
                    "id": f"{_ORDER_PREFIX}{oid}",
                    "title": title or f"Order {oid}",
                    "url": _ui_order_url(client, oid),
                }
            )

        # Parties.
        parties_resp = await client.get(
            "parties",
            params=odata_params(top=10, full_text_search=query, orderby="Name asc"),
        )
        for p in _items(parties_resp):
            pid = p.get("PartyID") or p.get("Id")
            if pid is None:
                continue
            results.append(
                {
                    "id": f"{_PARTY_PREFIX}{pid}",
                    "title": p.get("Name") or f"Party {pid}",
                    "url": _ui_party_url(client, pid),
                }
            )

        return {"results": results}

    @mcp.tool(
        annotations={
            "title": "Fetch Billit record",
            "readOnlyHint": True,
            "openWorldHint": True,
        }
    )
    async def fetch(id: str) -> dict[str, Any]:  # noqa: A002 — name fixed by spec
        """Fetch a single record by the ID returned from ``search``."""
        if id.startswith(_ORDER_PREFIX):
            oid = id[len(_ORDER_PREFIX) :]
            order = await client.get(f"orders/{oid}")
            return {
                "id": id,
                "title": (
                    f"{order.get('OrderType', 'Order')} {order.get('OrderNumber', '')}"
                ).strip(),
                "text": json.dumps(order, ensure_ascii=False, indent=2, default=str),
                "url": _ui_order_url(client, oid),
                "metadata": {
                    "type": order.get("OrderType"),
                    "direction": order.get("OrderDirection"),
                    "status": order.get("OrderStatus"),
                    "totalIncl": order.get("TotalIncl"),
                    "paid": order.get("Paid"),
                },
            }
        if id.startswith(_PARTY_PREFIX):
            pid = id[len(_PARTY_PREFIX) :]
            party = await client.get(f"parties/{pid}")
            return {
                "id": id,
                "title": party.get("Name") or f"Party {pid}",
                "text": json.dumps(party, ensure_ascii=False, indent=2, default=str),
                "url": _ui_party_url(client, pid),
                "metadata": {
                    "type": party.get("PartyType"),
                    "vat": party.get("VATNumber"),
                    "email": party.get("Email"),
                },
            }
        return {
            "id": id,
            "title": "Unknown record",
            "text": (
                f"Unknown ID format {id!r}. Expected '{_ORDER_PREFIX}<id>' "
                f"or '{_PARTY_PREFIX}<id>'."
            ),
            "url": "",
        }


def _items(resp: Any) -> list[dict[str, Any]]:
    if isinstance(resp, dict):
        items = resp.get("Items")
        if isinstance(items, list):
            return items
    if isinstance(resp, list):
        return resp
    return []


def _ui_order_url(client: BillitClient, oid: object) -> str:
    base = client.settings.normalized_base_url
    ui = base.replace("api.", "my.").replace("/v1", "")
    return f"{ui}/Order/Edit/{oid}"


def _ui_party_url(client: BillitClient, pid: object) -> str:
    base = client.settings.normalized_base_url
    ui = base.replace("api.", "my.").replace("/v1", "")
    return f"{ui}/Party/Edit/{pid}"

File: node/src/tools/searchFetch.ts
/**
 * ChatGPT Apps SDK compatibility — `search` and `fetch` with the exact result
 * shapes OpenAI's connectors / Deep Research expect:
 *
 *   search(query) -> { results: [{ id, title, url }] }
 *   fetch(id)     -> { id, title, text, url, metadata? }
 *
 * Both are mapped over Billit orders + parties. Responses include
 * structuredContent plus a JSON text mirror, per the Apps SDK guidance.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import type { BillitClient } from "../client.js";
import { odataParams } from "./util.js";
import { BillitError } from "../errors.js";

const ORDER_PREFIX = "order:";
const PARTY_PREFIX = "party:";

type Rec = Record<string, unknown>;

function items(resp: unknown): Rec[] {
  if (Array.isArray(resp)) return resp as Rec[];
  const inner = (resp as Rec | null)?.Items;
  return Array.isArray(inner) ? (inner as Rec[]) : [];
}

function uiBase(client: BillitClient): string {
  return client.config.baseUrl.replace("api.", "my.");
}

function structured(data: Rec) {
  return {
    structuredContent: data,
    content: [{ type: "text" as const, text: JSON.stringify(data, null, 2) }],
  };
}

function errorResult(err: unknown) {
  const text =
    err instanceof BillitError
      ? err.toDisplay()
      : err instanceof Error
        ? `${err.name}: ${err.message}`
        : String(err);
  return { isError: true, content: [{ type: "text" as const, text }] };
}

export function register(server: McpServer, client: BillitClient): void {
  server.registerTool(
    "search",
    {
      title: "Search Billit",
      description:
        "Search invoices/credit notes/offers and customers/suppliers across Billit. " +
        "Returns result records with ids usable by the fetch tool.",
      inputSchema: { query: z.string().describe("Free-text search query.") },
      outputSchema: {
        results: z.array(
          z.object({ id: z.string(), title: z.string(), url: z.string() }),
        ),
      },
      annotations: { readOnlyHint: true, openWorldHint: true },
    },
    async ({ query }) => {
      try {
        const results: { id: string; title: string; url: string }[] = [];

        const orders = await client.get("orders", {
          params: odataParams({ top: 10, fullTextSearch: query, orderby: "OrderDate desc" }),
        });
        for (const o of items(orders)) {
          const oid = o.OrderID ?? o.Id;
          if (oid === undefined || oid === null) continue;
          const customer = (o.Customer as Rec | undefined)?.Name ?? "";
          const title = `${o.OrderType ?? "Order"} ${o.OrderNumber ?? ""} — ${customer}`
            .replace(/ — $/, "")
            .trim();
          results.push({
            id: `${ORDER_PREFIX}${oid}`,
            title: title || `Order ${oid}`,
            url: `${uiBase(client)}/Order/Edit/${oid}`,
          });
        }

        const parties = await client.get("parties", {
          params: odataParams({ top: 10, fullTextSearch: query, orderby: "Name asc" }),
        });
        for (const p of items(parties)) {
          const pid = p.PartyID ?? p.Id;
          if (pid === undefined || pid === null) continue;
          results.push({
            id: `${PARTY_PREFIX}${pid}`,
            title: (p.Name as string) || `Party ${pid}`,
            url: `${uiBase(client)}/Party/Edit/${pid}`,
          });
        }

        return structured({ results });
      } catch (err) {
        return errorResult(err);
      }
    },
  );

  server.registerTool(
    "fetch",
    {
      title: "Fetch Billit record",
      description: "Fetch a single record by an id returned from the search tool.",
      inputSchema: {
        id: z.string().describe("Record id from search, e.g. 'order:123' or 'party:45'."),
      },
      outputSchema: {
        id: z.string(),
        title: z.string(),
        text: z.string(),
        url: z.string(),
        metadata: z.record(z.string(), z.unknown()).optional(),
      },
      annotations: { readOnlyHint: true, openWorldHint: true },
    },
    async ({ id }) => {
      try {
        if (id.startsWith(ORDER_PREFIX)) {
          const oid = id.slice(ORDER_PREFIX.length);
          const order = (await client.get(`orders/${oid}`)) as Rec;
          return structured({
            id,
            title: `${order.OrderType ?? "Order"} ${order.OrderNumber ?? ""}`.trim(),
            text: JSON.stringify(order, null, 2),
            url: `${uiBase(client)}/Order/Edit/${oid}`,
            metadata: {
              type: order.OrderType,
              direction: order.OrderDirection,
              status: order.OrderStatus,
              totalIncl: order.TotalIncl,
              paid: order.Paid,
            },
          });
        }
        if (id.startsWith(PARTY_PREFIX)) {
          const pid = id.slice(PARTY_PREFIX.length);
          const party = (await client.get(`parties/${pid}`)) as Rec;
          return structured({
            id,
            title: (party.Name as string) || `Party ${pid}`,
            text: JSON.stringify(party, null, 2),
            url: `${uiBase(client)}/Party/Edit/${pid}`,
            metadata: {
              type: party.PartyType,
              vat: party.VATNumber,
              email: party.Email,
            },
          });
        }
        return structured({
          id,
          title: "Unknown record",
          text: `Unknown id format '${id}'. Expected '${ORDER_PREFIX}<id>' or '${PARTY_PREFIX}<id>'.`,
          url: "",
        });
      } catch (err) {
        return errorResult(err);
      }
    },
  );
}
